src/utils/resources.py

```diff
@@ -257,6 +258,143 @@ def _run_script(self, action, script, env={}):
+            properties[source_id].update(infos)
+
+        super().__init__({"sources": properties}, *args, **kwargs)
+
+    def deprovision(self, context: Dict = {}):
+        if os.path.isdir(f"/var/cache/yunohost/download/{self.app}/"):
+            rm(f"/var/cache/yunohost/download/{self.app}/", recursive=True)
+
+    def provision_or_update(self, context: Dict = {}):
+
+        # Don't prefetch stuff during restore
+        if context.get("action") == "restore":
+            return
+
+        for source_id, infos in self.sources.items():
+
+            if not infos["prefetch"]:
+                continue
+
+            if infos["url"] is None:
+                arch = system_arch()
+                if arch in infos and isinstance(infos[arch], dict) and isinstance(infos[arch].get("url"), str) and isinstance(infos[arch].get("sha256"), str):
+                    self.prefetch(source_id, infos[arch]["url"], infos[arch]["sha256"])
+                else:
+                    raise YunohostError(f"In resources.sources: it looks like you forgot to define url/sha256 or {arch}.url/{arch}.sha256", raw_msg=True)
+            else:
+                if infos["sha256"] is None:
+                    raise YunohostError(f"In resources.sources: it looks like the sha256 is missing for {source_id}", raw_msg=True)
+                self.prefetch(source_id, infos["url"], infos["sha256"])
+
+    def prefetch(self, source_id, url, expected_sha256):
+
+        logger.debug(f"Prefetching asset {source_id}: {url} ...")
+
+        if not os.path.isdir(f"/var/cache/yunohost/download/{self.app}/"):
+            mkdir(f"/var/cache/yunohost/download/{self.app}/", parents=True)
+        filename = f"/var/cache/yunohost/download/{self.app}/{source_id}"
+
+        # NB: we use wget and not requests.get() because we want to output to a file (ie avoid ending up with the full archive in RAM)
+        # AND the nice --tries, --no-dns-cache, --timeout options ...
+        p = subprocess.Popen(["/usr/bin/wget", "--tries=3", "--no-dns-cache", "--timeout=900", "--no-verbose", "--output-document=" + filename, url], stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        out, _ = p.communicate()
+        returncode = p.returncode
+        if returncode != 0:
+            if os.path.exists(filename):
+                rm(filename)
+            out = out.decode()
+            raise YunohostError(f"Failed to download asset {source_id} ({url}) for {self.app}: {out}", raw_msg=True)
+
+        assert os.path.exists(filename), f"For some reason, wget worked but {filename} doesnt exists?"
+
+        computed_sha256 = check_output(f"sha256sum {filename}").split()[0]
+        if computed_sha256 != expected_sha256:
+            size = check_output(f"du -hs {filename}").split()[0]
+            rm(filename)
+            raise YunohostError(f"Corrupt source for {url} : expected to find {expected_sha256} as sha256sum, but got {computed_sha256} instead ... (file size : {size})", raw_msg=True)
```

Comment:
Don't we want to translate all these `YunohostError`?

Reply:
Zblerg yeah I was a bit lazy, but yeah, at least the most "probable" ones, such as this one

Reply:
Done :+1: